includes/wc-template-functions.php

```diff
@@ -1830,46 +1830,69 @@ function woocommerce_maybe_show_product_subcategories( $loop_html ) {
 	}
 }
 
-if ( ! function_exists( 'woocommerce_get_product_subcategories' ) ) {
+if ( ! function_exists( 'woocommerce_product_subcategories' ) ) {
 	/**
-	 * Get (and cache) product subcategories.
+	 * This is a legacy function which used to check if we needed to display subcats and then output them. It was called by templates.
 	 *
-	 * @param int $parent_id Get subcategories of this ID.
-	 * @return array
-	 */
-	function woocommerce_get_product_subcategories( $parent_id = 0 ) {
-		$parent_id          = absint( $parent_id );
-		$product_categories = wp_cache_get( 'product-categories-' . $parent_id, 'product_cat' );
+	 * From 3.3 onwards this is all handled via hooks and the woocommerce_maybe_show_product_subcategories function.
+	 *
+	 * Since some templates have not updated compatibility, to avoid showing incorrect categories this function has been deprecated and will
+	 * return nothing. Replace usage with woocommerce_output_product_categories to render the category list manually.
+	 *
+	 * This is a legacy function which also checks if things should display.
+	 * Themes no longer need to call these functions. It's all done via hooks.
+	 *
+	 * @deprecated 3.3.1
+	 * @param array $args Arguments.
+	 * @return null|boolean
+	*/
+	function woocommerce_product_subcategories( $args = array() ) {
+		$defaults = array(
+			'before'        => '',
+			'after'         => '',
+			'force_display' => false,
+		);
 
-		if ( false === $product_categories ) {
-			// NOTE: using child_of instead of parent - this is not ideal but due to a WP bug ( https://core.trac.wordpress.org/ticket/15626 ) pad_counts won't work.
-			$product_categories = get_categories( apply_filters( 'woocommerce_product_subcategories_args', array(
-				'parent'       => $parent_id,
-				'menu_order'   => 'ASC',
-				'hide_empty'   => 0,
-				'hierarchical' => 1,
-				'taxonomy'     => 'product_cat',
-				'pad_counts'   => 1,
-			) ) );
-			wp_cache_set( 'product-categories-' . $parent_id, $product_categories, 'product_cat' );
-		}
+		$args = wp_parse_args( $args, $defaults );
 
-		if ( apply_filters( 'woocommerce_product_subcategories_hide_empty', true ) ) {
-			$product_categories = wp_list_filter( $product_categories, array( 'count' => 0 ), 'NOT' );
-		}
+		if ( $args['force_display'] ) {
```

Comment:
Why even have this option? :)

If someone is editing their theme to add the special `force_display` arg it's the same amount of work for them as changing the function call to`woocommerce_output_product_categories`.

If you do keep this option, I think the deprecation notice should show when using it. Right now it only shows when not using the force option.

That's all just nitpicking, though.  This works well and I think it's a good solution.